### PR TITLE
[SCR-138] feat: Add RequestInterceptor

### PR DESCRIPTION
**Pull request by doubleface**
Allows the konnectors to wait for any http request interception and more + a documentation

### CODE REVIEW

**Review by KillianCourvoisier — APPROVED** [review]

File: packages/cozy-clisk/docs/RequestInterceptor.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    url: 'test/api',
+    serialization: 'json'
+  }
+])
+requestInterceptor.init()
+
+const connector = new ContentScript({ requestInterceptor })
+```
+
+This will intercept all requests with method 'GET' and containing 'test/api' in their url. The
```
Comment (Crash--) [inline packages/cozy-clisk/docs/RequestInterceptor.md:26]: so this is `*test/api*` right?

File: packages/cozy-clisk/docs/RequestInterceptor.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+response  of these requests will be serialized to a json object.
+
+## Fetching the response
+
+You can fetch the intercepted responses in two ways : `waitForRequestInterception` or directly
```
Comment (Crash--) [inline packages/cozy-clisk/docs/RequestInterceptor.md:31]: response

File: packages/cozy-clisk/docs/RequestInterceptor.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  ContentScript,
+  RequestInterceptor
+} from 'cozy-clisk/dist/contentscript'
+
+const requestInterceptor = new RequestInterceptor([
+  {
+    label: 'testRequest',
+    method: 'GET',
+    url: 'test/api',
+    serialization: 'json'
```
Comment (Crash--) [inline packages/cozy-clisk/docs/RequestInterceptor.md:18]: What is it? Is it the `content/type` ? the accepted header?

File: packages/cozy-clisk/src/contentscript/ContentScript.js [hunk 4]
```diff
@@ -165,8 +182,22 @@ export default class ContentScript {
...
 
+    if (!this.bridge) {
+      throw new Error(
+        'No bridge is defined, you should call ContentScript.init before using this method'
+      )
+    }
+
     if (contentScriptType === WORKER_TYPE) {
+      this.requestInterceptor?.on('response', response => {
+        this.bridge?.emit('workerEvent', {
```
[the remark this answers is not in the document]
Comment (Crash--) [reply packages/cozy-clisk/src/contentscript/ContentScript.js:193]: Isn't weird to add this listener before calling workerready ?

File: packages/cozy-clisk/src/contentscript/ContentScript.js [hunk 5]
```diff
@@ -268,6 +299,44 @@ export default class ContentScript {
...
+    const timeout = options?.timeout ?? 60000
+
+    const interceptionPromise = new Promise(resolve => {
+      const listener = ({ event, payload }) => {
+        if (event === 'requestResponse' && payload.label === label) {
+          if (!this.bridge) {
+            throw new Error(
+              'No bridge is defined, you should call ContentScript.init before using this method'
+            )
+          }
```
Comment (Crash--) [inline packages/cozy-clisk/src/contentscript/ContentScript.js:321]: no need to add this test, you already added it before adding the listener

File: packages/cozy-clisk/src/contentscript/ContentScript.spec.js [hunk 6]
```diff
@@ -5,11 +5,11 @@ describe('ContentScript', () => {
   describe('saveFiles', () => {
     it('should call launcher saveFiles with given nominal options', async () => {
       const contentScript = new ContentScript()
-      contentScript.setContentScriptType(PILOT_TYPE)
       contentScript.bridge = {
         call: jest.fn(),
         emit: jest.fn()
       }
+      contentScript.contentScriptType = PILOT_TYPE
```
Comment (Crash--) [inline packages/cozy-clisk/src/contentscript/ContentScript.spec.js:12]: why this change?

File: packages/cozy-clisk/docs/RequestInterceptor.md (hunk 1 above)
Comment (doubleface) [reply packages/cozy-clisk/docs/RequestInterceptor.md:26]: I chose to make these stars implicit, you think we should accept glob patterns or regexp ?

File: packages/cozy-clisk/docs/RequestInterceptor.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+```javascript
+import {
+  ContentScript,
+  RequestInterceptor
+} from 'cozy-clisk/dist/contentscript'
+
+const requestInterceptor = new RequestInterceptor([
+  {
+    label: 'testRequest',
```
[the remark this answers is not in the document]
Comment (doubleface) [reply packages/cozy-clisk/docs/RequestInterceptor.md:15]: Right, I think identifier will be better

File: packages/cozy-clisk/docs/RequestInterceptor.md (hunk 3 above)
Comment (doubleface) [reply packages/cozy-clisk/docs/RequestInterceptor.md:18]: This is how we choose to serialize the body of the response, since it needs to be serialized to be transmitted from the worker to the pilot. We could relate it to the content type but I preferred it to be explicit.

File: packages/cozy-clisk/src/contentscript/ContentScript.js (hunk 5 above)
Comment (doubleface) [reply packages/cozy-clisk/src/contentscript/ContentScript.js:321]: This was actually ts-check which enforces me to do it. In theory, nothing avoid to remove the bridge between the two tests :shrug:

File: packages/cozy-clisk/src/contentscript/ContentScript.spec.js (hunk 6 above)
Comment (doubleface) [reply packages/cozy-clisk/src/contentscript/ContentScript.spec.js:12]: To avoid the call to setContentScript in this test, since setContentScript does more now.

File: packages/cozy-clisk/src/contentscript/ContentScript.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (doubleface) [reply packages/cozy-clisk/src/contentscript/ContentScript.js:193]: 1) typechecking too
2) right ! The konnector should be able to subscribe to these event before on event is emitted. I will fix it

File: packages/cozy-clisk/src/contentscript/ContentScript.js [hunk 8]
```diff
@@ -165,8 +182,22 @@ export default class ContentScript {
...
+    if (!this.bridge) {
+      throw new Error(
+        'No bridge is defined, you should call ContentScript.init before using this method'
+      )
+    }
+
     if (contentScriptType === WORKER_TYPE) {
+      this.requestInterceptor?.on('response', response => {
+        this.bridge?.emit('workerEvent', {
+          event: 'requestResponse',
```
[the remark this answers is not in the document]
Comment (doubleface) [reply packages/cozy-clisk/src/contentscript/ContentScript.js:194]: Since these are events already coming from RequestInterceptor, I found it redundant, no ?

File: packages/cozy-clisk/docs/RequestInterceptor.md (hunk 1 above)
Comment (Crash--) [reply packages/cozy-clisk/docs/RequestInterceptor.md:26]: I don't like the implicit, and I think we should avoid it. I don't know what we should accept, but regex seems a good candidate.

File: packages/cozy-clisk/src/contentscript/ContentScript.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Crash--) [reply packages/cozy-clisk/src/contentscript/ContentScript.js:194]: When you have a large codebase and you start looking for something, I can assume you that there is nothing redundant. Because looking for `requestResponse` will get you a lot of responses. And the majority will not be related to the ContentScript. So you'll have to look for `requestInterceptor` and then you'll find a few places where it's used. And then you'll have to read those places to look for `requestResponse`. 

I know that is this the idea of the grep test (https://jamie-wong.com/2013/07/12/grep-test/) taken to the extreme, but this is how I can navigate into the all cozy's codebase.

File: packages/cozy-clisk/src/contentscript/ContentScript.spec.js (hunk 6 above)
Comment (Crash--) [reply packages/cozy-clisk/src/contentscript/ContentScript.spec.js:12]: What else does it do?
